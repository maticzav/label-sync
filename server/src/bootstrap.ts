import { compile } from 'handlebars'
import { GHTree } from './github'
import { mapEntries } from './data/dict'

/**
 * Populates the template with repositories.
 * @param tree
 */
export function populateTempalte(
  tree: GHTree,
  data: { repositories: { name: string }[] },
): GHTree {
  return mapEntries(tree, file => compile(file)(data))
}
