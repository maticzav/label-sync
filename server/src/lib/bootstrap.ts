import * as handlebars from 'handlebars'
import * as path from 'path'
import * as prettier from 'prettier'

import { mapEntries } from '../data/dict'
import { GHTree } from './github'

/**
 * Populates the template with repositories.
 * @param tree
 */
export function populateTemplate(tree: GHTree, data: { repository: string; repositories: { name: string }[] }): GHTree {
  return mapEntries(tree, (file, name) => {
    /* Personalize file */
    const populatedFile = handlebars.compile(file)(data)

    /* Format it */
    switch (path.extname(name)) {
      case '.ts': {
        return prettier.format(populatedFile, { parser: 'typescript' })
      }
      case '.yml': {
        return prettier.format(populatedFile, { parser: 'yaml' })
      }
      case '.md': {
        return prettier.format(populatedFile, { parser: 'markdown' })
      }
      case '.json': {
        return prettier.format(populatedFile, { parser: 'json' })
      }
      default: {
        return populatedFile
      }
    }
  })
}
