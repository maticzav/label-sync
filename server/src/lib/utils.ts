/**
 * Negates the wrapped function.
 * @param fn
 */
export function not<T, L extends Array<T>>(fn: (...args: L) => boolean): (...args: L) => boolean {
  return (...args) => !fn(...args)
}

/**
 * Creates a fallback default value.
 * @param fallback
 * @param value
 */
export function withDefault<T>(fallback: T, value: T | undefined | null): T {
  if (value == null) return fallback
  return value
}
