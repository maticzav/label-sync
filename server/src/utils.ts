import fs from 'fs'
import path from 'path'
import { mapKeys } from './data/dict'

/**
 * Negates the wrapped function.
 * @param fn
 */
export function not<T, L extends Array<T>>(
  fn: (...args: L) => boolean,
): (...args: L) => boolean {
  return (...args) => !fn(...args)
}

/**
 * Loads a tree of utf-8 decoded files at paths.
 *
 * @param path
 */
export function loadTreeFromPath(
  root: string,
  ignore: (string | RegExp)[],
): { [path: string]: string } {
  const files = fs.readdirSync(root, { encoding: 'utf-8' })
  const tree = files
    .filter(file => !ignore.some(glob => RegExp(glob).test(file)))
    .flatMap(file => {
      const rootFilePath = path.resolve(root, file)
      if (fs.lstatSync(rootFilePath).isDirectory()) {
        return Object.entries(
          mapKeys(loadTreeFromPath(rootFilePath, ignore), key =>
            unshift(file, key),
          ),
        )
      } else {
        return [[file, fs.readFileSync(rootFilePath, { encoding: 'utf-8' })]]
      }
    })

  return Object.fromEntries(tree)
}

/**
 * Adds a folder to the path.
 * @param pre
 * @param path
 */
function unshift(pre: string, path: string): string {
  return [pre, ...path.split('/').filter(Boolean)].join('/')
}
