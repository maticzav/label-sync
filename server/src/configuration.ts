import * as t from 'io-ts'
import { PathReporter } from 'io-ts/lib/PathReporter'
import yaml from 'js-yaml'
import * as os from 'os'

import { mapEntries } from './data/dict'

/**
 * Configuration repository is the repository which LabelSync
 * uses to determine the configuration of the service.
 *
 * @param organization
 */
export const getLSConfigRepoName = (owner: string) => `${owner}-labelsync`

/**
 * Configuration file path determines the path of the file in the repositoy
 * that we use to gather configuration information from.
 *
 * It should always be in YAML format.
 */
export const LS_CONFIG_PATH = 'labelsync.yml'

/**
 * Label represents the central unit of LabelSync. Each label
 * can have multiple siblings that are meaningfully related to
 * a label itself, multiple hooks that trigger different actions.
 */
const LSCLabel = t.intersection([
  t.type({
    color: t.string,
  }),
  t.partial({
    description: t.string,
    siblings: t.array(t.string),
    alias: t.array(t.string),
  }),
])
export type LSCLabel = t.TypeOf<typeof LSCLabel>

const LSCLabelName = t.string
export type LSCLabelName = t.TypeOf<typeof LSCLabelName>

/**
 * Repository configuration for how LabelSync should sync it.
 */
const LSCRepositoryConfiguration = t.partial({
  removeUnconfiguredLabels: t.boolean,
})
export interface LSCRepositoryConfiguration
  extends t.TypeOf<typeof LSCRepositoryConfiguration> {}

/**
 * Repository represents a single Github repository.
 * When configured as `strict` it will delete any surplus of labels
 * in the repository.
 */
const LSCRepository = t.intersection([
  t.partial({
    config: LSCRepositoryConfiguration,
  }),
  t.type({
    labels: t.record(LSCLabelName, LSCLabel),
  }),
])
export interface LSCRepository extends t.TypeOf<typeof LSCRepository> {}
const LSCRepositoryName = t.string
export type LSCRepositoryName = t.TypeOf<typeof LSCRepositoryName>

/**
 * Configuration represents an entire configuration for all
 * LabelSync tools that an organisation is using.
 */
const LSCConfiguration = t.type({
  repos: t.record(LSCRepositoryName, LSCRepository),
})
export interface LSCConfiguration extends t.TypeOf<typeof LSCConfiguration> {}

/**
 * Parses a string into a configuration object.
 * @param input
 */
export function parseConfig(
  input: string,
  logger?: (err: any) => any,
): [string] | [null, LSCConfiguration] {
  try {
    const object = yaml.safeLoad(input)
    const config = LSCConfiguration.decode(object)

    if (config._tag === 'Left') {
      return [PathReporter.report(config).join(os.EOL)]
    }
    return [null, fixConfig(config.right)]
  } catch (err) /* istanbul ignore next */ {
    if (logger) logger(err)
    return [err.message]
  }
}

/**
 * Catches configuration issues that io-ts couldn't catch.
 *
 * @param config
 */
function fixConfig(config: LSCConfiguration): LSCConfiguration {
  const repos = mapEntries(config.repos, fixRepoConfig)

  return {
    ...config,
    repos: repos,
  }
}

/**
 * Fixes issues that io-ts couldn't catch.
 *
 * @param config
 */
function fixRepoConfig(config: LSCRepository): LSCRepository {
  const labels = mapEntries<LSCLabel, LSCLabel>(config.labels, label => ({
    ...label,
    color: fixLabelColor(label.color),
  }))

  return {
    ...config,
    labels,
  }
}

/**
 * Removes hash from the color.
 *
 * @param color
 */
function fixLabelColor(color: string): string {
  if (color.startsWith('#')) return color.slice(1)
  return color
}
