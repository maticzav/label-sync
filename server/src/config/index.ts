import { Plan } from '@prisma/client'

import * as parser from './parser'
import { checkConfiguration } from './checks/configuration'
import { checkPlan } from './checks/plan'
import { getLSConfigRepoName } from './constants'

import * as t from './types'

/**
 * This file bundles up all functions related to configuration.
 */

export * from './constants'
export * from './types'

/**
 * Parses the configuration.
 */
export function parse(params: {
  plan: Plan
  input: string
}): [string] | [null, t.LSCConfiguration] {
  const result = parser.parse(params.input, [
    checkPlan(params.plan),
    checkConfiguration,
  ])

  if (result.success) {
    return [null, result.config]
  }
  return [result.error]
}

// MARK: - Utility functions

/**
 * Returns a list of repositories in configuration without wildcard config.
 */
export function configRepos(config: t.LSCConfiguration): string[] {
  let repos: string[] = []

  for (let repo of config.repos.keys()) {
    repo = repo.trim()
    if (repo !== '*') repos.push(repo)
  }
  return repos
}

/**
 * Tells whether the repository is the configuration
 * repository for the account.
 */
export function isConfigRepo(account: string, repo: string): boolean {
  return getLSConfigRepoName(account) === repo.toLowerCase()
}
