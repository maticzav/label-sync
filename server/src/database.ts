/**
 * This file contains the datasource that connects to the database.
 */

import { PrismaClient } from '@prisma/client'
import * as prisma from '@prisma/client'

import * as t from 'luxon'

import * as c from './config'
import { Maybe } from './data/maybe'

export class Data {
  // MARK: - State

  private prisma: PrismaClient

  // MARK: - Constructor

  constructor(prisma: PrismaClient) {
    this.prisma = prisma
  }

  // MARK: - Methods

  /**
   * Returns the configuration of a particular installation.
   */
  async config(owner: string, repo: string): Promise<Maybe<c.LSCRepository>> {
    const [config] = await this.prisma.repositoryConfiguration.findMany({
      where: {
        configuration: { active: { owner: owner } },
        repository: { in: [repo, '*'] },
      },
      orderBy: {
        repository: 'asc',
      },
      include: {
        labels: true,
      },
    })

    if (config === undefined) return null

    let configuration: c.LSCRepository = {
      config: {
        removeUnconfiguredLabels: config.removeUnconfiguredLabels,
      },
      labels: new Map(),
    }

    for (const label of config.labels) {
      configuration.labels.set(label.name, {
        color: label.color,
        description: label.description,
        alias: label.alias,
        siblings: label.siblings,
      })
    }

    return configuration
  }

  /**
   * Returns the user plan.
   */
  async plan(owner: string): Promise<prisma.Plan> {
    const now = t.DateTime.now()

    const installation = await this.prisma.installation.findUnique({
      where: { owner: owner },
    })

    /**
     * Check that we have an installation and that plan is still valid.
     */
    if (installation === null) return 'FREE'
    if (t.DateTime.fromJSDate(installation.periodEndsAt) < now) return 'FREE'

    return installation.plan
  }

  async install(): Promise<void> {}
}
