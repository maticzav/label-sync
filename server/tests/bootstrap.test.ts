import fs from 'fs'
import path from 'path'

import { populateTempalte } from '../src/bootstrap'
import { loadTreeFromPath } from '../src/utils'

const templatesPath = path.resolve(__dirname, '../../templates')
const templates = fs.readdirSync(templatesPath).map((template) => ({
  name: template,
  path: path.resolve(templatesPath, template),
}))

describe('templates:', () => {
  const data = {
    repository: 'maticzav-labelsync',
    repositories: [
      {
        name: 'graphql-shield',
      },
      {
        name: 'label-sync',
      },
      {
        name: 'resk',
      },
    ],
  }

  for (const template of templates) {
    test(`populates ${template.name} tempalate`, () => {
      const tree = loadTreeFromPath(template.path, [
        'dist',
        'node_modules',
        '.DS_Store',
        /.*\.log.*/,
        /.*\.lock.*/,
      ])

      expect(populateTempalte(tree, data)).toMatchSnapshot()
    })
  }
})
