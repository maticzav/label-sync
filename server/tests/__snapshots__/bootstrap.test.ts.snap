// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates: populates yaml tempalate 1`] = `
Object {
  ".gitignore": "dist
node_modules

.DS_Store
*.log*",
  ".huskyrc.json": "{
  \\"hooks\\": {
    \\"pre-commit\\": \\"npm run make\\"
  }
}
",
  "README.md": "# LabelSync Template

Hi :wave:,

Welcome to LabelSync. We are very excited to see you use our service. We believe that managing organisations should be simple, that's why we built LabelSync.

## What this template includes?

- \`label-sync\`: utility package for generating configuration file
- \`huskyrc\`: pre-commit hooks
- \`prettier\`: code formatting
",
  "labelsync.yml": "repos: {}
",
  "package.json": "{
  \\"name\\": \\"labelsync\\",
  \\"version\\": \\"0.0.0\\",
  \\"private\\": true,
  \\"main\\": \\"index.js\\",
  \\"scripts\\": {
    \\"make\\": \\"ts-node src/index.ts\\"
  },
  \\"dependencies\\": {
    \\"husky\\": \\"3.1.0\\",
    \\"label-sync\\": \\"latest\\"
  },
  \\"devDependencies\\": {
    \\"prettier\\": \\"1.19.1\\",
    \\"ts-node\\": \\"8.6.2\\",
    \\"typescript\\": \\"3.8.3\\"
  },
  \\"license\\": \\"MIT\\"
}
",
  "prettier.config.js": "module.exports = {
  semi: false,
  singleQuote: true,
  trailingComma: 'all',
}
",
  "src/colors.ts": "export const colors = {
  area: 'FFD700',
  kind: '3B5BDB',
  status: 'F8F9FA',
  bug: 'ee0000',
  priority: 'F783AC',
  scope: '27CF79',
  team: 'FDF4E8',
  release: 'a5d8ff',
  process: 'EB9100',
}
",
  "src/index.ts": "import { configuration, make } from 'label-sync'

import { github } from './repositories/github'
import { prisma } from './repositories/prisma'

/**
 * Create a configuration instance by definining repositories
 * and their configurations.
 */
const config = configuration({
  repositories: {
    /**
     * replace repository name with your own repositories
     */
    // 'emma-cli': github,
  },
})

/* Generates labelsync.yml into the root of repository. */
make({
  configs: [config],
})
",
  "src/repositories/github.ts": "import { repository, label } from 'label-sync'

/**
 * Default collection of label in a Github repository.
 */
export const github = repository({
  strict: false,
  labels: {
    bug: label({
      color: '#d73a4a',
      description: \\"Something isn't working\\",
    }),
    documentation: label({
      color: '#0075ca',
      description: 'Improvements or additions to documentation',
    }),
    duplicate: label({
      color: '#cfd3d7',
      description: 'This issue or pull request already exists',
    }),
    enhancement: label({
      color: '#a2eeef',
      description: 'New feature or request',
    }),
    'good first issue': label({
      color: '#7057ff',
      description: 'Good for newcomers',
    }),
    'help wanted': label({
      color: '#008672',
      description: 'Extra attention is needed',
    }),
    invalid: label({
      color: '#e4e669',
      description: \\"This doesn't seem right\\",
    }),
    question: label({
      color: '#d876e3',
      description: 'Further information is requested',
    }),
    wontfix: label({
      color: '#000000',
      description: 'This will not be worked on',
    }),
  },
})
",
  "src/repositories/prisma.ts": "import { repository, label } from 'label-sync'

import { colors } from '../colors'

/**
 * Label configuration used internally by Prisma team. Labels are grouped
 * by their intention (e.g. bug/*, kind/*, process/*) and give
 * great base for issue triaging.
 */

export const prisma = repository({
  strict: false,
  labels: {
    /* Bugs */
    'bug/0-needs-info': label({
      color: colors.bug,
      description: 'More information is needed for reproduction.',
    }),
    'bug/1-repro-available': label({
      color: colors.bug,
      description: 'A reproduction exists and needs to be confirmed.',
    }),
    'bug/2-confirmed': label({
      color: colors.bug,
      description: 'We have confirmed that this is a bug.',
    }),
    /* Kind */
    'kind/bug': label({
      color: colors.kind,
      description: 'A reported bug.',
    }),
    'kind/regression': label({
      color: colors.kind,
      description: 'A reported bug in functionality that used to work before.',
    }),
    'kind/feature': label({
      color: colors.kind,
      description: 'A request for a new feature.',
    }),
    'kind/improvement': label({
      color: colors.kind,
      description: 'An improvement to existing feature and code.',
    }),
    'kind/docs': label({
      color: colors.kind,
      description: 'A documentation change is required.',
    }),
    'kind/discussion': label({
      color: colors.kind,
      description: 'Discussion is required.',
    }),
    'kind/question': label({
      color: colors.kind,
      description: 'Developer asked a question. No code changes required.',
    }),
    /* Process triaging. */
    'process/candidate': label({
      color: colors.process,
      description: 'Candidate for next Milestone.',
    }),
    'process/next-milestone': label({
      color: colors.process,
      description: 'Issue earmarked for next Milestone.',
    }),
    'process/product': label({
      color: colors.process,
      description:
        'Temporary label to export products issues from the Engineering process',
    }),
  },
})
",
  "tsconfig.json": "{
  \\"compilerOptions\\": {
    \\"target\\": \\"es2017\\",
    \\"module\\": \\"commonjs\\",
    \\"lib\\": [\\"dom\\", \\"es2017\\", \\"es2019\\"],
    \\"rootDir\\": \\"src\\",
    \\"outDir\\": \\"dist\\"
  }
}
",
}
`;

exports[`templates: populates yaml tempalate 2`] = `
Object {
  ".gitignore": "dist
node_modules

.DS_Store
*.log*",
  "README.md": "",
  "labelsync.yml": "# Colors
colors:
  area: &area '#FFD700'
  kind: &kind '#3B5BDB'
  status: &status '#F8F9FA'
  bug: &bug '#EE0000'
  priority: &priority '#F783AC'
  scope: &scope '#27CF79'
  team: &team'#FDF4E8'
  release: &release '#A5D8ff'
  process: &process' '#EB9100'

# Labels
labels: &common
  'bug/0-needs-info':
    color: &bug
    description: More information is needed for reproduction.
  'bug/1-repro-available':
    color: &bug
    description: A reproduction exists and needs to be confirmed.
  'bug/2-confirmed':
    color: &bug
    description: We have confirmed that this is a bug.
  'kind/bug':
    color: &bug
    description: A reported bug.
  'kind/regression':
    color: &kind
    description: A reported bug in functionality that used to work before.
  'kind/feature':
    color: &kind
    description: A request for a new feature.
  'kind/improvement':
    color: &kind
    description: An improvement to existing feature and code.
  'kind/docs':
    color: &kind
    description: A documentation change is required.
  'kind/discussion':
    color: &kind
    description: Discussion is required.
  'kind/question':
    color: &kind
    description: Developer asked a question. No code changes required.
  'process/candidate':
    color: &process
    description: Candidate for next Milestone.
  'process/next-milestone':
    color: &process
    description: Issue earmarked for next Milestone.
  'process/product':
    color: &process
    description: Temporary label to export products issues from the Engineering process

repos:
    
    
",
}
`;
